homeassistant/components/shelly/switch.py

```diff
@@ -144,3 +155,46 @@ async def async_turn_on(self, **kwargs: Any) -> None:
     async def async_turn_off(self, **kwargs: Any) -> None:
         """Turn off relay."""
         await self.call_rpc("Switch.Set", {"id": self._id, "on": False})
+
+
+class ShellyOtaUpdateBetaChannelSwitch(SwitchEntity):
+    """Defines a Shelly OTA update beta channel switch."""
+
+    _attr_icon = "mdi:flask-outline"
+    _attr_entity_category = ENTITY_CATEGORY_CONFIG
+
+    def __init__(
+        self, wrapper: RpcDeviceWrapper | BlockDeviceWrapper, entry: ConfigEntry
+    ) -> None:
+        """Initialize Shelly OTA update beta channel switch."""
+        self._attr_device_info = DeviceInfo(
+            connections={(CONNECTION_NETWORK_MAC, wrapper.mac)}
+        )
+
+        if isinstance(wrapper, RpcDeviceWrapper):
+            device_name = get_rpc_device_name(wrapper.device)
+        else:
+            device_name = get_block_device_name(wrapper.device)
+
+        self._attr_name = f"{device_name} OTA Update Beta channel"
+        self._attr_unique_id = slugify(self._attr_name)
+
+        self.entry = entry
+        self.wrapper = wrapper
+
+    @property
+    def is_on(self) -> bool:
+        """Return true if device is on."""
+        return bool(self.entry.options.get(CONF_OTA_BETA_CHANNEL))
+
+    async def async_turn_on(self, **kwargs: Any) -> None:
+        """Turn the device on."""
+        self.hass.config_entries.async_update_entry(
+            self.entry, options={**self.entry.options, CONF_OTA_BETA_CHANNEL: True}
+        )
+
+    async def async_turn_off(self, **kwargs: Any) -> None:
+        """Turn the device off."""
+        self.hass.config_entries.async_update_entry(
+            self.entry, options={**self.entry.options, CONF_OTA_BETA_CHANNEL: False}
+        )
```

Comment:
I think this needs to be discussed.

This is basically making "virtual" entities, as this switch is not provided by the device. Additionally, it "misuses" (not the right term, as it needs to be discussed), the integration entry options to store the device configuration for this virtual entity.

IMHO, the latter is wrong (as this is not an integration option, but a device option), however, we do lack a concept for this at this moment as well. (e.g., a single integration entry providing multiple devices cannot benefit from this code here).

The questions then become:

> Do we even want such a concept?, if so, how would we handle this nicely for all situations?

Reply:
to be honest, to get this PR ready, i would remove the switch and implement an options-flow.
This is suitable here, since each device has its own config entry.

About the device option discussion, I'm very interested in and i am willing to push this forward, since i have already another use case, where one config entry creates multiple device, but to have an option for each device independently would be great (_use case would be setting the snapshot quality per camera in Synology DSM integration - see #58306_)
I assume there are many other use cases out there, but also guess this is better placed in an architectural discussion?